README.md
VEXTREME Vectrex Multicart
===

We're carrying on and pushing forward the amazing work of [Sprite_tm](http://spritesmods.com/?art=veccart&page=1) and his Extreme Vectrex Multicart. He posted the code (GPLv3 License) and lots of screens of the PCB/initial schematic, but he never released the PCB. It's a great base for an inexpensive and open multicart, for developers and players alike!

:warning: Please read this NOTICE of Current Development :warning:
===

**This project is a work in progress, and is not ready for mass production.**  Please observe the currently [posted issues](https://github.com/technobly/vextreme/issues) and there may be more lurking that are not yet identified.  The hardware and software is evolving.  You are free to build any version, contribute and/or follow along.  Please note that support of these pre-releases will be very limited and if you embark on an epic adventure to build one or a few, you should feel comfortable with this type of embedded hardware and software development.  There will be road bumps, but the journey is half the fun of the destination.  When things appear to be stable and we've worked out most of the kinks, we'll change the version to v1.0.  Only then is it advisable to mass produce.

3D Renders
===

![v0.2 Render](images/vextreme-v0.2-white.gif)

Video Updates
===

Be sure to subscribe to my Youtube channel for updates!  Here's my [Vectrex playlist](https://www.youtube.com/watch?v=zkJ-z77fJCw&list=PL7MgXfpGKg6CCaIQejZVb5CsTg3-5mHU0&index=1) if you want to stay laser focused on this project.

VEXTREME Discord Server
===
Join the [VEXTREME Discord server](https://discord.gg/VDssGVJ) to chat with us about the development

BOM and Parts ordering
===

All the parts are described in [veccart.csv](bom/veccart.csv)

Also, you can use this [Digi-Key shared cart](https://www.digikey.com/short/zpdmtp) if you're in a hurry.  It has will have part references right on the packages for you!

Ordering PCB's
===

OSHPark is a good place to order with purple or the new "after dark" theme color scheme.  You can upload the KiCad [veccart.kicad_pcb](hardware/veccart.kicad_pcb) there directly.  I would download this entire Github repo ZIP file first though instead of just trying to save the PCB file from your browser.

Another way to order PCB's is by using the included [gerbers](hardware/gerbers) and uploading those with all of the necessary specs to companies like [PCBWay](https://www.pcbway.com) or [JLCPCB](https://jlcpcb.com)

Building and Flashing STM firmware
===

### This assumes you have these prerequisits installed:

- Docker
- dfu-util v0.9
- Make
- Windows will also require these [Zadig instructions](https://github.com/profezzorn/ProffieOS/wiki/zadig)

### Build the stm32-build docker image
```
code/veccart $ make docker-build
```

### Connect the 2-pin jumper and then connect the cart to USB

### Check if your computer sees the STM32
```
$ dfu-util -l

dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2019 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Deducing device DFU version from functional descriptor length
Found DFU: [0483:df11] ver=2200, devnum=22, cfg=1, intf=0, path="20-1.4.3", alt=3, name="@Device Feature/0xFFFF0000/01*004 e", serial="123412341234"
Found DFU: [0483:df11] ver=2200, devnum=22, cfg=1, intf=0, path="20-1.4.3", alt=2, name="@OTP Memory /0x1FFF7800/01*512 e,01*016 e", serial="123412341234"
Found DFU: [0483:df11] ver=2200, devnum=22, cfg=1, intf=0, path="20-1.4.3", alt=1, name="@Option Bytes  /0x1FFFC000/01*016 e", serial="123412341234"
Found DFU: [0483:df11] ver=2200, devnum=22, cfg=1, intf=0, path="20-1.4.3", alt=0, name="@Internal Flash  /0x08000000/04*016Kg,01*064Kg,03*128Kg", serial="123412341234"
```

### Build and flash the STM32 image via dfu-util
```
code/veccart $ make clean all flash USE_HW=v0.3

// NOTE: Use v0.2 above if you have v0.2 HW, or you may use v0.3 if you make the proper mods to your board.  See "Modifying v0.2 to v0.3"

// You should end up with something like this

Downloading to address = 0x08000000, size = 23536
Download    [=========================] 100%        23536 bytes
Download done.
File downloaded successfully
```

### Remove the cart from USB and remove the jumper (install it on one pin so you don't lose it)


Building and Flashing the multicart's menu
===

### Build the asm6809 docker image
```
code/multicart $ make docker-build
```

### Connect the cart to USB with the 2-pin jumper removed

- The first time you do this you will likely have to format the USB drive

  - Name: VEXTREME
  - Format: MS-DOS (FAT)
  - Scheme: Master Boot Record
  - 512 byte page size

- Be patient, it's slow!  Don't worry, copying binaries later will be fast!!

### Build and flash the multicart menu binary
```
// Mac OS
code/multicart $ make clean all && cp multicart.bin /Volumes/VEXTREME/

// optionally add this to the above command to unmount, after you figure out which drive it is
$ diskutil list
&& diskutil unmountDisk /dev/disk3

// Linux
code/multicart $ make copy

// You should end up with something like this, and the binary should have been copied to the root of your multicart drive

asm6809  -B -o multicart.bin multicart.asm
```

Modifying v0.2 to v0.3
===

![v0.3 Mod 1](images/v0.3-mod1.png)

- To get the new reset feature of v0.3, as of v0.24 software you will need to jumper V-OE pin 12 of the cart fingers (or U3 pin 9) to STM32 pin 29 (PB10).  This may change in the future, don't sell your soldering iron!
- The wire mod above also allows you to play Animaction on VEXTREME!
- More TBD


Adding Vectrex Game ROMs (Binaries)
===

- Create a directory on your multicart called `roms/`

- Add your first binaries there, may I suggest: [Beluga Dreams](https://8-bit-waves.itch.io/beluga-dreams) and [Vec-Man](http://eiti.fh-pforzheim.de/personen/johannsen/projektlabor/vectrex_2019/vectrex_game_vec_man.htm) ?

Docs
===

A nice collection of lessons learned will be [documented here.](docs/)


Hall of Fame
===

If you build one of these, be sure to send me a message with your build pics showing your working VEXTREME multicart!  Or better yet, check out the [hall of fame format here](hall-of-fame/) and submit a PR (please and thank you!)

LICENSE
===

GPLv3 - essentially you must keep this open, all changes must be disclosed and shared.  Full [LICENSE here](LICENSE) and [TL;DR summary here](https://tldrlegal.com/license/gnu-general-public-license-v3-(gpl-3))


CHANGELOG.md
# Changelog

## Hardware TODO

- Move v0.2's PCB fingers 18.0mm from the center of 3.5mm mounting holes, and 0.5mm from board edge, for slightly better engagement with the Vectrex cartrdige connector.
- Remove soldermask from PCB cart fingers on top and bottom of the board.
- R4 & D3 reference designators are swapped.
- Remove D4 - D6 and just have addressible LEDs on the edge of the board, space the 7 LEDs evenly.
- Move the debugging/serial port connector so it does not interfere with enclosure bosses.
- Move V-HALT and V-CE to different pins and fix SDAT and SCLK so they are on SPI2 for hardware control. V-CE might be optimized for MSB (seems like it is in romemu.S), so move to PA15 which bumps CSn to some other pin? CSn should likely be on SPI1_SS, but it might not matter (check).  PB01 is used in an optimized way as well though.
- Replace U3 with a single OR gate?
- Replacement for U4 which is more available at LCSC?
- Slightly increase the middle pads on addressible LEDs so they don't require touch up soldering
- Route all unused GPIO to a breakout header, for hacking!
- Add ESD protection to external connections
- Move headers to USB-C side and edge for easier access
- Make 2-pin jumper a right angle switch instead of a jumper?

## Software TODO

- Add hardware SPI2 support
- Add cartridge RAM support (specifically for Animaction $2000-$27FF, but possibly for others as well).  Any downside to letting the entire 64KB of emulated ROM to be considered as RAM and writable?
- Add hardware version support for pin definitions
- Add 128KB bankswitching (V-INT pin)
- Developer mode, allow the cart to stay plugged into a computer and appear as a RAMDISK

## v0.23

- Fixes issue where VEXTREME takes several minutes to mount as USB drive on Windows #53

## v0.22

- ignore all file extensions that are not .bin or .vec (case insensitive) #49
- increased quick reset delay to 700ms #45 
- adds HW/SW version to VEXTREME menu and games #48 
  - see example in LED Test v1.1.bin

## v0.21

- :warning: Added NOTICE of Current Development
- Fixes Polar Rescue and implements a new Reset Heurisic #37
- Adds addressable LED control and demo #41
    - added 3 functions to control the LEDs from a Vectrex program
        - updateOne() - RPC 6
        - updateAll() - RPC 4
        - updateMulti() - RPC 7
    - added RPC function 5 to step through rainbow control of LEDs in menu
    - added "LED TEST" which uses updateMulti to show how to finely control
    all of the LEDs with a test program.  Press UP on the joystick, try it slowly as well :D
    - added addressable LED control to Veccy Bird v1.5

## v0.2

- Replaced USB-B-mini with USB-C and centered USB on cart
- Extended the PCB by 1.7 mm to get the USB-C connector as close to the case exterior as possible.
- Widened the PCB to 48.0 mm so that there’s less side to side play in the cartridge slot.
- Moved outer mounting holes into the proper locations
- Reversed D3 and R4 order to get D3 closer to edge of the PCB, but it turned out we added something else there!
- Added 10 APA102-2020 addressable RGB LED lights (max. draw 245mA) … qualify with USB-IN (PA9) for 100% brightness, else 50%
- Added 47uF cap for peak currents required for RGB LEDs.
- Added V-IRQ connection to PB9 for 128KB bank-switching
- Grounded unused floating inputs on U3 to reduce current consumption
- Adjusted Y1 to use 12pF load crystal resonator with 18pF load caps (see equation)
- Lots of clean up and tweaking all over PCB
- Added revision to PCB
- Added “VEXTREME” to PCB
- Added a ground pour under the STM32
- Made the contacts on the edge connector thinner (1.52mm) and made sure they were spaced properly (2.54mm).
- Started moving footprints to the libs/veccart.pretty library so you know where they will be!
- Add basic folder support #1
- Ignore dotfiles generated by MacOS
- Add software driven addressible LED support
- improve USB write time for a single game by 3x (32KB in 1 second)
- adds 'clean' and 'all' rules for multicart build
- create 4KB of space for menu (currently about 2KB), and reserve 60KB for menu data.
- Added a mostly optimized 100 vector logo for VEXTREME
- Added a much improved version of Malban's 5 line font (everything except lowercase, don't use lower)
- Reduced menu to 4 items at a time
- Added a special intro jingle that's been stuck in my head (enjoy!)
- temp increase time to reset to allow Fortress of Narzod to play
- Extract menu code from main code and optimize menu code
- Add basic menu labels/function-per-button support #4
    - At the bottom of the menu 1 = (BACK), 2 = PAGE LEFT, 3 = PAGE RIGHT, 4 = SELECT
- Ignore lst files #5
- adds ISSUE_TEMPLATE.md and PULL_REQUEST_TEMPLATE.md for contributions from the community
- Update to latest (pinned) libopencm3 #6
- Fixed lastselectcart on wrong game after reset
- Fixed do not inc page if next one will be blank issue
- Converts logo from sync to smart list, thanks Malban!
    - Note: Had to replace NOP MACROs with JSR of approx equivalent value
    - Converted fcb hi(label),lo(label) into fdb label
    - Reduces cycle count by 8 - 10k!

## v0.1

- Mostly unchanged from Sprite_tm's original HW/SW design
- Mostly unchanged from Rattboi's original PCB design
- fixes error opening PCB with latest KiCaD
- add gerbers for easy PCB ordering at many places
- update README.md and add LICENSE
- fixes premature automatic reset delay on GCE Vectrex
- fixes menu builder missing filename in path
- pretty up the menu by removing the .BIN file extension, and other cleanup
- let's call it the VEXTREME cart, for now. (original name was EXTREME)


code/veccart/led.h
/*------------------------------------------------------------------------
  Ported to STM32F411 for VEXTREME by Technobly

  ------------------------------------------------------------------------
  -- original header follows ---------------------------------------------
  ------------------------------------------------------------------------

  This file is part of the Adafruit Dot Star library.

  Adafruit Dot Star is free software: you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public License
  as published by the Free Software Foundation, either version 3 of
  the License, or (at your option) any later version.

  Adafruit Dot Star is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with NeoPixel.  If not, see <http://www.gnu.org/licenses/>.
  ------------------------------------------------------------------------*/

#ifndef _ADAFRUIT_DOT_STAR_H_
#define _ADAFRUIT_DOT_STAR_H_

// Color-order flag for LED pixels (optional extra parameter to constructor):
// Bits 0,1 = R index (0-2), bits 2,3 = G index, bits 4,5 = B index
#define RGB_RGB (0 | (1 << 2) | (2 << 4))
#define RGB_RBG (0 | (2 << 2) | (1 << 4))
#define RGB_GRB (1 | (0 << 2) | (2 << 4))
#define RGB_GBR (2 | (0 << 2) | (1 << 4))
#define RGB_BRG (1 | (2 << 2) | (0 << 4))
#define RGB_BGR (2 | (1 << 2) | (0 << 4))

// Addressable LED Preset Colors
static uint32_t colors[] = {
    0,        // off
    0xFF0000, // red
    0xFF9900, // orange
    0xFFFF00, // yellow
    0x00FF00, // green
    0x00FFFF, // cyan
    0x0000FF, // blue
    0x7700FF, // pink
    0xFF00FF, // magenta
    0xFFFFFF  // white
};

/* 8-bit gamma-correction table.
   Copy & paste this snippet into a Python REPL to regenerate:
import math
gamma=2.6
for x in range(256):
    print("{:3},".format(int(math.pow((x)/255.0,gamma)*255.0+0.5))),
    if x&15 == 15: print
*/
static const uint8_t _LedsGammaTable[256] = {
    0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,
    0,  0,  0,  0,  0,  0,  0,  0,  1,  1,  1,  1,  1,  1,  1,  1,
    1,  1,  1,  1,  2,  2,  2,  2,  2,  2,  2,  2,  3,  3,  3,  3,
    3,  3,  4,  4,  4,  4,  5,  5,  5,  5,  5,  6,  6,  6,  6,  7,
    7,  7,  8,  8,  8,  9,  9,  9, 10, 10, 10, 11, 11, 11, 12, 12,
   13, 13, 13, 14, 14, 15, 15, 16, 16, 17, 17, 18, 18, 19, 19, 20,
   20, 21, 21, 22, 22, 23, 24, 24, 25, 25, 26, 27, 27, 28, 29, 29,
   30, 31, 31, 32, 33, 34, 34, 35, 36, 37, 38, 38, 39, 40, 41, 42,
   42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57,
   58, 59, 60, 61, 62, 63, 64, 65, 66, 68, 69, 70, 71, 72, 73, 75,
   76, 77, 78, 80, 81, 82, 84, 85, 86, 88, 89, 90, 92, 93, 94, 96,
   97, 99,100,102,103,105,106,108,109,111,112,114,115,117,119,120,
  122,124,125,127,129,130,132,134,136,137,139,141,143,145,146,148,
  150,152,154,156,158,160,162,164,166,168,170,172,174,176,178,180,
  182,184,186,188,191,193,195,197,199,202,204,206,209,211,213,215,
  218,220,223,225,227,230,232,235,237,240,242,245,247,250,252,255};

// Fun routines
uint32_t Wheel(uint8_t WheelPos);
void rainbow(uint8_t wait);
void rainbowCycle(uint8_t wait);
void colorWipe(bool dir, uint32_t c, uint8_t wait);
uint32_t dimColor(uint32_t color, uint8_t width);
uint32_t colorWheel(uint8_t WheelPos);
void knightRider(uint16_t cycles, uint16_t speed, uint8_t width, uint16_t first, uint16_t last, uint32_t color);
void theaterChaseRainbow(int wait);

void ledsInitHW(uint16_t n, uint8_t o);
void ledsInitSW(uint16_t n, uint32_t dp, uint16_t d, uint32_t cp, uint16_t c, uint8_t o);
void updateLength(uint16_t n);
void ledsEnd(void);                                  // Destructor
void ledsBegin(void);                                // Prime pins/SPI for output
void ledsClear(void);                                // Set all pixel data to zero
void ledsSetBrightness(uint8_t);                     // Set global brightness 0-255
void ledsSetPixelColor(uint16_t n, uint32_t c);
void ledsSetPixelColorRGB(uint16_t n, uint8_t r, uint8_t g, uint8_t b);
void ledsUpdate(void);                               // Issue color data to strip

uint32_t ledsColor(uint8_t r, uint8_t g, uint8_t b); // R,G,B to 32-bit color
uint32_t ledsGetPixelColor(uint16_t n);              // Return 32-bit pixel color

uint16_t ledsNumPixels(void);                        // Return number of pixels

uint8_t ledsGetBrightness(void);                     // Return global brightness
uint8_t *ledsGetPixels(void);                        // Return pixel data pointer

/**
 * A gamma-correction function for 32-bit packed RGB or WRGB
 * colors. Makes color transitions appear more perceptially correct.
 */
uint32_t ledsColorHSV2(uint16_t hue, uint8_t sat, uint8_t val);
static uint32_t ledsColorHSV(uint32_t x) {
    return ledsColorHSV2(x, 255, 255); // sat = 255, val = 255
}

static uint8_t ledsGamma8(uint8_t x) {
    return _LedsGammaTable[x]; // 0-255 in, 0-255 out
}

uint32_t ledsGamma32(uint32_t x);

#endif // _ADAFRUIT_DOT_STAR_H_
